### DISCUSSION TITLE
[RFC] Caching clickhouse queries?

### DISCUSSION THREAD

**Discussion by TheBuilderJR** [plausible/analytics#1681] [Anything Else]
I'd like to start contributing the project and thought of potentially one first idea I can contribute. According to Clickhouse docs, https://clickhouse.com/docs/en/introduction/performance/, we ideally shouldn't send more than 100 QPS to clickhouse

> Since this scenario is not typical for analytical DBMSs, we recommend expecting a maximum of 100 queries per second.

I wonder if it makes sense to have some sort of in memory LRU cache to reduce the amount of scanning we do for clickhouse reads? We could set a TTL of one minute to begin with. At the very least enabling this for embeds (for instance on https://plausible.io/plausible.io) seems like a good way to improve efficiency.

**Comment by ukutaht** [comment]
Clickhouse already maintains a LRU cache for queries as far as I know. Introducing a higher-level cache in the Elixir app would cut some requests and bandwidth for sure.

I've thought about this a bit deeper and honestly I don't think the Elixir app is the place with highest leverage in terms of caching. I would start from the frontend app instead.

Assuming that the data is immutable, all the historical data can be cached in the frontend and requesting it never needs to go the app server or the database. If I'm looking at the 'Last 12 months' view for the live demo, the last 11 months (which have ended) should be requested from the server just once. After the first load, I should never need to ask the server what happened 11 months ago again, the stats should be cached on the frontend.

BTW this only applies to unfiltered data. When filters are applied, we should always hit the database.

This is a very fundamental shift because the frontend needs to become aware of times and dates. At the moment we do minimal date transformations in the frontend but if we want to cache, the frontend needs to be able to slice and dice data/queries by dates in a flexible way.

This would be helpful for frontend caching: https://react-query.tanstack.com/
